bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/module/handler/GenericEventConditionHandler.java

```diff
@@ -27,50 +30,69 @@
-    private static final String SOURCE = "source";
-    private static final String PAYLOAD = "payload";
+    public static final String CFG_TOPIC = "topic";
+    public static final String CFG_TYPES = "types";
+    public static final String CFG_SOURCE = "source";
+    public static final String CFG_PAYLOAD = "payload";
 
     public final Logger logger = LoggerFactory.getLogger(GenericEventConditionHandler.class);
 
+    private final String source;
+    private final @Nullable TopicGlobEventFilter topicFilter;
+    private final Set<String> types;
+    private final @Nullable Pattern payloadPattern;
+
     public GenericEventConditionHandler(Condition module) {
         super(module);
-    }
 
-    private boolean isConfiguredAndMatches(String keyParam, @Nullable String value) {
-        Object mo = module.getConfiguration().get(keyParam);
-        String configValue = mo != null && mo instanceof String ? (String) mo : null;
-        if (configValue != null) {
-            if (PAYLOAD.equals(keyParam)) {
-                // automatically adding wildcards only for payload matching
-                configValue = configValue.startsWith("*") ? configValue : ".*" + configValue;
-                configValue = configValue.endsWith("*") ? configValue : configValue + ".*";
-            }
-            if (value != null) {
-                return value.matches(configValue);
-            } else {
-                return false;
-            }
+        this.source = (String) module.getConfiguration().get(CFG_SOURCE);
+        String topic = (String) module.getConfiguration().get(CFG_TOPIC);
+        if (!topic.isEmpty()) {
+            topicFilter = new TopicGlobEventFilter(topic);
+        } else {
+            topicFilter = null;
+        }
+        if (module.getConfiguration().get(CFG_TOPIC) != null) {
+            this.types = Set.of(((String) module.getConfiguration().get(CFG_TOPIC)).split(","));
+        } else {
+            this.types = Set.of();
+        }
+        String payload = (String) module.getConfiguration().get(CFG_PAYLOAD);
+        if (!payload.isEmpty()) {
+            payloadPattern = Pattern.compile(payload);
         } else {
-            return true;
+            payloadPattern = null;
         }
     }
 
     @Override
     public boolean isSatisfied(Map<String, Object> inputs) {
         Event event = inputs.get("event") != null ? (Event) inputs.get("event") : null;
-        if (event != null) {
-            return isConfiguredAndMatches(TOPIC, event.getTopic()) && isConfiguredAndMatches(SOURCE, event.getSource())
-                    && isConfiguredAndMatches(PAYLOAD, event.getPayload())
-                    && isConfiguredAndMatches(EVENTTYPE, event.getType());
+        if (event == null) {
+            return false;
         }
-        return false;
+        if (!types.isEmpty() && !types.contains(event.getType())) {
+            return false;
+        }
+        TopicGlobEventFilter localTopicFilter = topicFilter;
```

Comment:
I was think about that in your previous PR. Wouldn't it be better to have a `MatchAllEventFilter` that always returns `true`? That way we could omit a lot of null checks and also make `.getFilter` return a non-null value.

In general since `EventFilter` gets the `Event`, a filter could also filter not only on topic but also on source or type. That brings me again to the `EventFilterBuilder` that builds an `EventFilter` `.withTopicGlob(String glob)`, `.withTopicRegex(String regex)`, `.withType(Class<? extends Event>)`, `.withTypes(Set<Class<? extends Event>>` or `.withSource(String source)`. The implementation than collects all those filters in a `List<EventFilter> filterChain`, and does a `filterChain.stream().allMatch(filter -> filter.apply(event)` in `.apply`. But this is out of scope here. Just an idea :-)

Reply:
I actually really like this idea. But yeah... out of scope here.